Microsoft.Identity.Web/Client/TokenCacheProviders/Session/MSALPerUserSessionTokenCacheProvider.cs

```diff
@@ -0,0 +1,233 @@
+            try
+            {
+                byte[] blob;
+                if (this.HttpContext.Session.TryGetValue(cacheKey, out blob))
+                {
+                    Debug.WriteLine($"INFO: Deserializing session {this.HttpContext.Session.Id}, cacheId {cacheKey}");
+                    this.UserTokenCache.DeserializeMsalV3(blob);
+                }
+                else
+                {
+                    Debug.WriteLine($"INFO: cacheId {cacheKey} not found in session {this.HttpContext.Session.Id}");
+                }
+            }
+            finally
+            {
+                SessionLock.ExitReadLock();
+            }
+        }
+
+        /// <summary>
+        /// Persists the user token blob to the Http session.
+        /// </summary>
+        private void PersistUserTokenCache()
+        {
+            string cacheKey = this.GetSignedInUsersUniqueId();
+
+            if (string.IsNullOrWhiteSpace(cacheKey))
+                return;
+
+            SessionLock.EnterWriteLock();
+
+            try
+            {
+                Debug.WriteLine($"INFO: Serializing session {this.HttpContext.Session.Id}, cacheId {cacheKey}");
+
+                // Reflect changes in the persistent store
+                byte[] blob = this.UserTokenCache.SerializeMsalV3();
+                this.HttpContext.Session.Set(cacheKey, blob);
+                this.HttpContext.Session.CommitAsync().Wait();
+            }
+            finally
+            {
+                SessionLock.ExitWriteLock();
+            }
+        }
+
+        /// <summary>
+        /// Clears the TokenCache's copy of this user's cache.
+        /// </summary>
+        public void Clear()
+        {
+            string cacheKey = this.GetSignedInUsersUniqueId();
+
+            if (string.IsNullOrWhiteSpace(cacheKey))
+                return;
+
+            SessionLock.EnterWriteLock();
+
+            try
+            {
+                Debug.WriteLine($"INFO: Clearing session {this.HttpContext.Session.Id}, cacheId {cacheKey}");
+
+                // Reflect changes in the persistent store
+                this.HttpContext.Session.Remove(cacheKey);
+                this.HttpContext.Session.CommitAsync().Wait();
+            }
+            finally
+            {
+                SessionLock.ExitWriteLock();
+            }
+
+            // Nulls the currently deserialized instance
+            this.LoadUserTokenCacheFromSession();
+        }
+
+        /// <summary>
+        /// if you want to ensure that no concurrent write take place, use this notification to place a lock on the entry
+        /// </summary>
+        /// <param name="args">Contains parameters used by the MSAL call accessing the cache.</param>
+
+        private void UserTokenCacheBeforeWriteNotification(TokenCacheNotificationArgs args)
+        {
+            // Since we obtain and release lock right before and after we read the Http session, we need not do anything here.
```

Comment:
Then we should not even subscribe?

Reply:
Because I wanted to provide guidance on why its not needed for individual providers

---
In reply to: [267390502](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/pull/65#discussion_r267390502) [](ancestors = 267390502)